docs/language-pack-providers/language-pack-contract.md

```diff
@@ -0,0 +1,94 @@
+The Knative Functions project provides the ability to customize its function templates and build strategies through Language Packs. Language Pack authors can add support for additional runtimes, provide templates that expose additional, custom built-in APIs, or any number of other capabilities. The function templates that are built into the `func` binary are considered the "built-in" or default Language Pack. The default, and all external Knative Function Language Packs should conform to the contract described in this document.
+
+## Language Pack Summary
+
+A Knative Function Language Pack provides runtime and invocation capabilities for user-provided Function code.
+
+- A Language Pack must be accessible as a git repository(???)
+- A Language Pack must provide utf8(plain?) text code template for function developers via `func create`
+- A Language Pack must expose an invokable function interface for function developers in the code template
+- A Language Pack must provide create, build, runtime and invocation metadata with a `manifest.yaml` file
+- A Language Pack project must be buildable in the form of an OCI container image via `func build`
+- A Language Pack OCI container image must be runnable via `func run`
+- A Language Pack project may be executable natively on a local host via host-specific tooling (e.g. `npm start`)
+
+## Components
+
+A Knative Function Language Pack consists, broadly, of two conceptual components.
+
+- Build and runtime metadata provided via its directory structure and, optionally, a `manifest.yaml` file, all of which supports the function lifecycle described below.
+- Project templates for functions and supporting code. This is the function developer's UX - a function project, which in most cases should look just like any other project of its type.
+
+## Build and Runtime Metadata
+
+// TODO(lance) inline this, or restructure the whole thing so that Lifecycle and this are all just one
+
+The build and runtime metadata specification for language packs can be found in the [Language Pack Guide](guides/language-packs.md).
+
+### Lifecycle
+
+- `create` Function projects are created using the `func create` command, during which, template and metadata files are copied from the Language Pack to a directory structure on the developer's local host.
+- `build` The function project is converted into a runnable OCI container image using the `func build` command with metadata provided by the Language Pack in `manifest.yaml`. Any dependencies declared by the function are installed onto the image file system, and the function invocation code is applied.
+- `run` Using the `func run` command to start the image, a controlling process loads the function project into memory and listens on port 8080 for incoming HTTP requests. The process is determined by the Language Pack. For example, a Node.js Language Pack may use `npm start` as the controlling process, while a Go Language Pack may invoke a binary compiled during the `build` phase.
```

Comment:
I think actually, that's something different. The `func` binary will choose another port on the host, but the function itself should always listen on 8080 which the binary will map to the chosen port (8080 or otherwise), and that is what is important for Language Pack providers to know here.